checked_message.go

```diff
@@ -41,6 +47,25 @@ type CheckedMessage struct {
 // NewCheckedMessage constructs a CheckedMessage. It's only intended for use by
 // wrapper libraries, and shouldn't be necessary in application code.
 func NewCheckedMessage(logger Logger, lvl Level, msg string) *CheckedMessage {
+	for {
+		val := cmPool.Get()
+		if val == nil {
+			break
+		}
+		m := val.(*CheckedMessage)
+		if n := m.used.Load(); n > 1 {
+			// we've already logged (and survived!) a DFatal log in the second
+			// CheckedMessage.Write, so just skip it and move on
+			continue
+		}
```

Comment:
I'm not sure that `used` is providing us the ironclad guarantees it used to - it's entirely possible to use a checked message once, keep a reference to it, and concurrently pull it out of the pool in another location. The second reference passes this test, since we've only used the message once, and then we hit line 61 and both callers see the new message. Now the original call site (which is misbehaving and re-using the checked message) is polluted with the message and level of the new call site. Both call sites have references to the same checked message, with a fresh atomic, so it's a race to see who gets to use it.

I don't know of a good pattern to avoid this sort of problem with pooling - it's inherently fiddly and error-prone, which is why we avoided it in the current implementation. I'm inclined to remove `used` altogether and make the documentation for `Write` and `Check` much, much more emphatic. (If you can think of a way to preserve a guarantee of safety, I'm all ears.)